### PR TITLE
UPSTREAM: 53606: use deployment pod template if 0 replicas

### PR DESCRIPTION
**Pull request by juanvallejo**
Fixes https://github.com/openshift/origin/issues/14286
Fixes https://github.com/openshift/origin/issues/14915

Prevent `oc debug` from hanging indefinitely when dealing
with a deployment that has been scaled down to 0.

cc @openshift/cli-review @smarterclayton

### CODE REVIEW

File: pkg/oc/cli/cmd/debug.go [hunk 1]
```diff
@@ -241,10 +242,20 @@ func (o *DebugOptions) Complete(cmd *cobra.Command, f *clientcmd.Factory, args [
 		return fmt.Errorf("you must identify a resource with a pod template to debug")
 	}
 
-	template, err := f.ApproximatePodTemplateForObject(infos[0].Object)
-	if err != nil && template == nil {
-		return fmt.Errorf("cannot debug %s: %v", infos[0].Name, err)
+	var template *kapi.PodTemplateSpec
```
Comment (smarterclayton) [inline pkg/oc/cli/cmd/debug.go]: Why wouldn't we always use ApproximatePodTemplateForObject for debug?

Comment (smarterclayton) [reply pkg/oc/cli/cmd/debug.go]: ApproximatePodTemplateForObject shouldn't wait for pods under DC

Comment (juanvallejo) [reply pkg/oc/cli/cmd/debug.go]: With a deployment object, the default case was being hit, as its type resolved to `*extensions.Deployment`. It then got stuck in the call to `f.AttachablePodForObject`. I could add a new case to the switch for `*extensions.Deployment` with the same logic as the `*deployapi.DeploymentConfig` case. wdyt?

Comment (juanvallejo) [reply pkg/oc/cli/cmd/debug.go]: https://github.com/openshift/origin/blob/master/pkg/cmd/util/clientcmd/factory.go#L209

Comment (smarterclayton) [reply pkg/oc/cli/cmd/debug.go]: There should be an "approximatepodtemplateforobject" for RCs and RSs and Deployments and StatefulSets and Daemonsets upstream that doesn't depend on pods existing.  So I would expect to always be using approximate pod template and for it to never block.

Comment (juanvallejo) [reply pkg/oc/cli/cmd/debug.go]: Without adding the case for `*extensions.Deployment`, current behavior hits the default case in https://github.com/openshift/origin/blob/master/pkg/cmd/util/clientcmd/factory.go#L209. Then `AttachablePodForObject` method is [called on factory_object_mapping upstream with a timeout of 1 min](https://github.com/openshift/origin/blob/master/pkg/cmd/util/clientcmd/factory.go#L210).

[Here is where the blocking seems to be happening](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/util/factory_object_mapping.go#L408).

Another possible solution could be to skip the call to `AttachablePodForObject` altogether if there are 0 replicas, and add a new type switch case for `*extensions.Deployment` [here](https://github.com/openshift/origin/blob/master/pkg/cmd/util/clientcmd/factory.go#L217) that just returns `t.Spec.Template`

Comment (smarterclayton) [reply pkg/oc/cli/cmd/debug.go]: Don't call AttachablePodForObject ever.  ApproximatePodTemplateforObject should always be used for debug all the way down, and nothing in APTFO should ever wait for a container to start (it doesn't have to).  attach requires a running pod, debug does not.

**Comment by juanvallejo** [comment]
@smarterclayton 
> There should be an "approximatepodtemplateforobject" for RCs and RSs and Deployments and StatefulSets and Daemonsets upstream that doesn't depend on pods existing. So I would expect to always be using approximate pod template and for it to never block.

Thanks, added `ApproximatePodTemplateForObject` to upstream factory to handle upstream types. Also removed use of `AttachablePodForObject` from this method downstream.

**Comment by juanvallejo** [comment]
ping @smarterclayton

**Comment by smarterclayton** [comment]
Test case for test-cmd that uses an RC scaled to zero and a DC scaled to zero to prove it works.

**Comment by juanvallejo** [comment]
@smarterclayton 
> Test case for test-cmd that uses an RC scaled to zero and a DC scaled to zero to prove it works.

Thanks, done

File: pkg/cmd/util/clientcmd/factory_object_mapping.go [hunk 2]
```diff
@@ -328,6 +330,70 @@ func (f *ring1Factory) StatusViewer(mapping *meta.RESTMapping) (kubectl.StatusVi
...
+		fallback = deployment.Spec.Template
+
+		pods, err := kc.Core().Pods(deployment.Namespace).List(metav1.ListOptions{LabelSelector: labels.SelectorFromSet(deployment.Spec.Selector).String()})
+		if err != nil {
+			return fallback, err
+		}
+
+		for i := range pods.Items {
+			pod := &pods.Items[i]
+			if fallback == nil || pod.CreationTimestamp.Before(fallback.CreationTimestamp) {
```
Comment (smarterclayton) [inline pkg/cmd/util/clientcmd/factory_object_mapping.go]: Why wouldn't you return the first pod no matter what, instead of the fallback?

Comment (smarterclayton) [reply pkg/cmd/util/clientcmd/factory_object_mapping.go]: Are you looking for pods newer than the most recent deployment?

Comment (juanvallejo) [reply pkg/cmd/util/clientcmd/factory_object_mapping.go]: @smarterclayton Yeah, this is existing code that I [moved from the clientcmd/factory](https://github.com/openshift/origin/blob/master/pkg/cmd/util/clientcmd/factory.go#L200) to clientcmd/factory_object_mapping. Can update the logic to just return the first pod if any exists instead if that is a better approach

**Comment by juanvallejo** [comment]
@smarterclayton friendly ping on https://github.com/openshift/origin/pull/16379#discussion_r142209198

File: pkg/cmd/util/clientcmd/factory_object_mapping.go (hunk 2 above)
Comment (smarterclayton) [reply pkg/cmd/util/clientcmd/factory_object_mapping.go]: Hrm, it's probably fine.  I just wasn't sure why it is here.  Add a comment to this section that describes the algorithm.

Comment (juanvallejo) [reply pkg/cmd/util/clientcmd/factory_object_mapping.go]: Thanks, done! Will go ahead and open PR for upstream changes
if everything else looks okay

**Comment by smarterclayton** [comment]
Has the upstream been turning into a PR?  Still says 00000

**Comment by juanvallejo** [comment]
@smarterclayton 
> Has the upstream been turning into a PR? Still says 00000

Opened a PR here: https://github.com/kubernetes/kubernetes/pull/53606
Commit has been reworded to include PR id

**Comment by smarterclayton** [comment]
Squash and lgtm

**Comment by juanvallejo** [comment]
Thanks, done!

**Comment by juanvallejo** [comment]
@smarterclayton [upstream] dep has merged, ptal

**Comment by juanvallejo** [comment]
@smarterclayton friendly ping - wondering if you could tag this for merging

**Pull request merged**